csi/src/dev/iscsi.rs

```diff
@@ -0,0 +1,249 @@
+use std::convert::TryFrom;
+
+use regex::Regex;
+use udev::Enumerator;
+use url::Url;
+use uuid::Uuid;
+
+use crate::dev::util::extract_uuid;
+
+use super::{Attach, Detach, DeviceError, DeviceName};
+
+mod iscsiadm;
+use iscsiadm::IscsiAdmin;
+
+pub(super) struct IscsiDevice {
+    portal: String,
+    iqn: String,
+    uuid: Uuid,
+    lun: u16,
+}
+
+impl IscsiDevice {
+    fn new(portal: String, iqn: String, uuid: Uuid, lun: u16) -> IscsiDevice {
+        IscsiDevice {
+            portal,
+            iqn,
+            uuid,
+            lun,
+        }
+    }
+
+    fn to_path(&self) -> String {
+        format!("ip-{}-iscsi-{}-lun-{}", self.portal, self.iqn, self.lun)
+    }
+
+    fn from_path(path: &str) -> Result<IscsiDevice, DeviceError> {
+        lazy_static! {
+            static ref PATTERN: Regex = Regex::new(r"^ip-(?P<host>[^:]+):(?P<port>[[:digit:]]+)-(?P<scheme>[[:alpha:]]+)-(?P<iqn>[[:alpha:]]+\.[[:digit:]]{4}-[[:digit:]]{2}\.[^:]+:(?P<suffix>[[:alnum:]]+(?:-[[:xdigit:]]+)+))-lun-(?P<lun>[[:digit:]]+)$").unwrap();
+        }
+
+        if let Some(captures) = PATTERN.captures(path) {
+            let host = captures.name("host").unwrap().as_str();
```

Comment:
I'm guessing these unwrap() won't fail because we've "captured" successfully?

Reply:
It is possible to create a regex with capture groups that are not used in a successful match (think disjunctions - ie. patterns of the form A|B). This can't happen in our pattern however, so yes it's safe to assume the capture groups all have Some() values. Even if the value of the group matches an empty string, this will still be a Some().